### ISSUE TITLE
"FATAL: sorry, too many clients already" postgres error in develop mode on 

### ISSUE THREAD

**Issue by macrozone** [prisma/prisma#1983]
# Bug description

I use prisma inside an api-route of https://nextjs.org/ (this is by the way an awesome setup) for a graphql-api. After some time you will get this error:

```

Error in connector: Error querying the database: db error: FATAL: sorry, too many clients already
    at PrismaClientFetcher.request
```

I guess the hot-reloading or refreshing of nextjs might mess with the connection-pool of prisma. I verified that the prisma-client that is used in the route is a singleton:

```
// this is imported in my api route
import { PrismaClient } from "@prisma/client";

const prisma = new PrismaClient();

export default prisma;
```

## How to reproduce

1. create a file pages/api/graphql.ts inside that, use prisma client
2. it probably needs some code changes that result in a rebuilding of this file /api/graphql.ts or its imports
3. at some point you should get the error

## Expected behavior

should not throw this error i guess?

## Environment & setup

- OS: macOS
- Database: [PostgreSQL
- Prisma version: prisma2@2.0.0-preview024, binary version: 377df4fe30aa992f13f1ba152cf83d5770bdbc85
- Node.js version: v12.13.1

**Comment by janpio** [comment]
Hey @macrozone, thanks for reporting this. Can you maybe create a minimal reproduction project with the API route so we do not mess up building the reproduction? Thanks.

**Comment by macrozone** [comment]
@janpio i think i can do that

**Comment by chynamyerz** [comment]
Hi

Any progress regarding this issue, have been encountering the same error often.

**Comment by janpio** [comment]
Can you create a minimal reproduction project @chynamyerz? For now we are still waiting on that before we can investigate and understand.

**Comment by macrozone** [comment]
sorry had no time yet to do a reproduction

**Comment by chynamyerz** [comment]
> Can you create a minimal reproduction project @chynamyerz? For now we are still waiting on that before we can investigate and understand.

Hi @janpio 

Yeah sure.

I tried to be as descriptive as possible but if I am clear as mud on the README please don't hesitate to contact me, will try to respond as quickly as possible.

Here is the link to the repo: https://github.com/chynamyerz/prisma2-too-many-clients-error.git

**Comment by cimchd** [comment]
We have the same problem in a similiar environment (prisma + nexus).
We already increased the max database connections to 200 and tested MariaDB and Postgres with the same problems.
It seems that the error occurs randomly. Sometimes directly on start, sometimes after several hours or even days...
We also connect to the database directly with a database management tool (HeidiSQL or DBeaver), if this may help...

Here the error message from prisma:
![image](https://user-images.githubusercontent.com/49520229/80182411-1dbc6380-8607-11ea-802f-03305e834372.png)
And here from the postgres database logs:
![image](https://user-images.githubusercontent.com/49520229/80182586-8acff900-8607-11ea-89fb-5122b41e74a3.png)

**Comment by pantharshit00** [comment]
@chynamyerz I looked at your code and you are creating a new PrismaClient instance per request. That will create a new connection pool per request so it will exhaust the database connection.

These lines of code are creating new client instance per request:
```js
const server = new GraphQLServer({ 
  context: ({ request }) => ({
    prisma: new PrismaClient(),
    request
  }),
  resolvers,
  typeDefs: __dirname + "/schema.graphql",
})
```

Instead, please invoke the client construct once and reuse the client instance like so:
```js
const prisma = new PrismaClient()

const server = new GraphQLServer({
  context: ({ request }) => ({
    prisma
    request,
  }),
  resolvers,
  typeDefs: __dirname + "/schema.graphql",
});
```

Anyone else subscribed to this issue, please make sure that your code is not creating a Prisma client instance per request.

**Comment by macrozone** [comment]
@pantharshit00 we use a single instance (we have `export const prisma = new PrismaClient()` in its own file which we import everywhere)

my guess is, that the hot-reloading / rebuilding of nextjs (or webpack in general) might rebuild and re-initialize this file. Is it possible to kindof destroy the instance (to free the connection pool)? then we could use maybe some re-build hook of webpack to reinitialize

**Comment by cimchd** [comment]
@pantharshit00 Our grapqhl code looks similiar to your example. So we should not have the problem at this point. But we habe several other occurences, where we have the two lines of code at the beginning of a file:
```js
const { PrismaClient } = require('@prisma/client');
const prisma = new PrismaClient();
```
I checked our code if we have it inside a function or some kind of iteration, but this is not the case. 
Should we refactor and also use only a single instance for the prisma client in our whole application?

**Comment by pantharshit00** [comment]
@cimchd You should indeed refactor it so that the code is using a single instance of the Prisma client. Every new call to `new PrismaClient()` will spin  up a new connection pool.


@macrozone You can try the following code to make sure next js hot reloading doesn't create a new prisma instance every hot reload:
```
import { PrismaClient } from "@prisma/client"

let prisma

if (process.env.NODE_ENV === "production") {
  prisma = new PrismaClient()
} else {
  if (!global.prisma) {
    global.prisma = new PrismaClient()
  }

  prisma = global.prisma
}

export default prisma
```


This snipped was originally posted here: https://github.com/prisma/prisma-client-js/issues/228#issuecomment-618433162

**Comment by macrozone** [comment]
@pantharshit00 thank you, will try out!

**Comment by chynamyerz** [comment]
> @chynamyerz I looked at your code and you are creating a new PrismaClient instance per request. That will create a new connection pool per request so it will exhaust the database connection.
> 
> These lines of code are creating new client instance per request:
> 
> ```js
> const server = new GraphQLServer({ 
>   context: ({ request }) => ({
>     prisma: new PrismaClient(),
>     request
>   }),
>   resolvers,
>   typeDefs: __dirname + "/schema.graphql",
> })
> ```
> 
> Instead, please invoke the client construct once and reuse the client instance like so:
> 
> ```js
> const prisma = new PrismaClient()
> 
> const server = new GraphQLServer({
>   context: ({ request }) => ({
>     prisma
>     request,
>   }),
>   resolvers,
>   typeDefs: __dirname + "/schema.graphql",
> });
> ```
> 
> Anyone else subscribed to this issue, please make sure that your code is not creating a Prisma client instance per request.

@pantharshit00 thank you, updated the code and it is now working just fine.

**Comment by jackHedaya** [comment]
Have Prisma exporting a singleton ever been considered?

**Comment by Sytten** [comment]
I dont think it would be wise since we extrapolate that one might be able to have multiple databases with different clients in the same project. Though it could use the same engine process under it.

**Comment by jackHedaya** [comment]
So in a case like that one would be able to create their own PrismaClient. However, it would be nice to have an a singleton as a simpler option for most cases.

**Comment by louisrli** [comment]
I solved my issue similarly -- in express I was instantiating one instance per request

Personally, I think this issue should be closed, in general whether it's a singleton or a new connection should be decidable by the user of the Prisma library.

Thanks for looking into this @pantharshit00

**Comment by macrozone** [comment]
Yes, its also solved for me. Still its worth get mentioned somewhere in the docs, as this is probably a common problem on some dev environement involving hot reloads

**Comment by giraffesyo** [comment]
For those having this issue, I solved it by creating a singleton:

```typescript
import { PrismaClient } from '@prisma/client'

class DBClient {
  public prisma: PrismaClient
  private static instance: DBClient
  private constructor() {
    this.prisma = new PrismaClient()
  }

  public static getInstance = () => {
    if (!DBClient.instance) {
      DBClient.instance = new DBClient()
    }
    return DBClient.instance
  }
}

export default DBClient
```

Then wherever I used the client I did:

```typescript
import DBClient from '../../database/client'
const prisma = DBClient.getInstance().prisma
```

You can then use it like normal.

**Comment by dilizarov** [comment]
@pantharshit00 I don't think the docs talk about recommending using a single instance of the prisma client.

I haven't used Prisma yet, but plan to in my next project and it was one of the key details I really wanted to know. Took this issue to find this information.

Perhaps the documentation should explicitly call it out?

**Comment by pantharshit00** [comment]
@dilizarov Opened: https://github.com/prisma/docs/issues/822

**Comment by juanzgc** [comment]
What exactly is the final answer to this? 

I followed this [example](https://github.com/prisma/prisma-examples/tree/latest/typescript/rest-nextjs-api-routes) however I am still seeing too many clients as I refresh a webpage and I am afraid this will cause errors in production.

**Comment by pantharshit00** [comment]
@juanzgc Docs issue already exists for this. It will be added.

**Comment by juanzgc** [comment]
> For those having this issue, I solved it by creating a singleton:
> 
> ```ts
> import { PrismaClient } from '@prisma/client'
> 
> class DBClient {
>   public prisma: PrismaClient
>   private static instance: DBClient
>   private constructor() {
>     this.prisma = new PrismaClient()
>   }
> 
>   public static getInstance = () => {
>     if (!DBClient.instance) {
>       DBClient.instance = new DBClient()
>     }
>     return DBClient.instance
>   }
> }
> 
> export default DBClient
> ```
> 
> Then wherever I used the client I did:
> 
> ```ts
> import DBClient from '../../database/client'
> const prisma = DBClient.getInstance().prisma
> ```
> 
> You can then use it like normal.

Just to note, after a day of using this I once again received: "Fatal: sorry, too many clients already"

**Comment by pantharshit00** [comment]
@juanzgc Where your database is hosted?

**Comment by chuckstock** [comment]
Having this same issue. We host our db on RDS with AWS. We see huge spikes in connections whenever we deploy additionally get this "Context creation failed: Error querying the database: db error: FATAL: sorry, too many clients already" pretty consistently.

**Comment by pantharshit00** [comment]
@chuckstock Can you please share how you are initializing and passing the client? Also, tell whether you are using prisma with next.js or not.

**Comment by chuckstock** [comment]
@pantharshit00 yes I'm using prisma with @nexus/schema, nextjs and graphql serverless api routes deployed on Vercel. 
```
import { PrismaClient } from '@prisma/client'

let prisma: PrismaClient

if (process.env.NODE_ENV === 'production') {
  if (!prisma) {
    prisma = new PrismaClient()
  }
} else {
  // @ts-ignore
  if (!global.prisma) {
    // @ts-ignore
    global.prisma = new PrismaClient()
  }

  // @ts-ignore
  prisma = global.prisma
}

export { prisma }
```

**Comment by juanzgc** [comment]
> @juanzgc Where your database is hosted?

It’s currently on my local machine. We have not pushed to production due to these issues.

**Comment by ColeTownsend** [comment]
Also have seen this!

**Comment by glenmcrae** [comment]
> Having this same issue. We host our db on RDS with AWS. We see huge spikes in connections whenever we deploy additionally get this "Context creation failed: Error querying the database: db error: FATAL: sorry, too many clients already" pretty consistently.

I think this is a separate issue - have a look at RDS proxy / PG Bouncer.

**Comment by pantharshit00** [comment]
Indeed, you will need to use a connection pooler if your serverless function is doing a lot of invocations so that you don't overload the connection quota. RDS Proxy/ PgBouncer are the two most commonly used connection poolers.

**Comment by tobkle** [comment]
Did the Singleton suggestion from above, but still receive that error on localhost for next.js and postgres.

**Comment by pantharshit00** [comment]
@tobkle Can you please share a reproduction with you setup?

**Comment by regirock365** [comment]
All the solutions mentioned above don't seem to fully solve the problem for me. As per the Prisma deployment guide, I set the prisma connection limit to 1 like so: `postgresql://USER:PASSWORD@HOST:PORT/DATABASE?connection_limit=1`. This reduced the frequency of the "too many clients" issue, but it still happens.

If I still want to keep the Next.js api route setup for Prisma, is there any solution that works?

**Comment by pantharshit00** [comment]
@regirock365 Do you mind sharing a reproduction in a new issue? This issue is getting hard to follow now so if anyone is having this issue, please open a new issue with a minimal reproduction.

**Comment by rajuashok** [comment]
Has anyone solved this for Typescript to make auto-complete work?

This is what I have:
```
// this is in /lib/prisma.ts
declare global {
    namespace NodeJS {
        interface Global {
            prisma: PrismaClient;
        }
    }
}

let prisma

if (process.env.NODE_ENV === "production") {
  prisma = new PrismaClient()
} else {
  if (!global.prisma) {
    global.prisma = new PrismaClient()
  }

  prisma = global.prisma
}

export default prisma
```

However, when I import prisma from this I no longer get auto-complete on `prisma.user` for example:
```
import prisma from '../../../lib/prisma';

const user = await prisma.use // <--- No auto-complete
```

**Comment by awey95** [comment]
This is a solution I came across, I don't see it posted here but hopefully it helps.

```typescript
import { PrismaClient } from '@prisma/client'

const prismaClientPropertyName = `__prevent-name-collision__prisma`
type GlobalThisWithPrismaClient = typeof globalThis & {
  [prismaClientPropertyName]: PrismaClient
}

const getPrismaClient = () => {
  if (process.env.NODE_ENV === `production`) {
    return new PrismaClient()
  } else {
    const newGlobalThis = globalThis as GlobalThisWithPrismaClient
    if (!newGlobalThis[prismaClientPropertyName]) {
      newGlobalThis[prismaClientPropertyName] = new PrismaClient()
    }
    return newGlobalThis[prismaClientPropertyName]
  }
}

const prisma = getPrismaClient()

export default prisma
```

@rajuashok this should solve your issue of not being able to get type-completion with the above approach.

**Comment by lccarrier** [comment]
I am having this issue locally using NextJS with Prisma/Nexus in API Routes. 

I wasn't able to solve it with solutions posted above.

**Comment by lccarrier** [comment]
I'm also having this problem using Next with Prisma/Nexus in API Routes.  I wasn't able to solve it with previous solutions.

**Comment by jackHedaya** [comment]
Give this a shot to those still dealing with the errors.

```ts
import { PrismaClient } from '@prisma/client'

declare global {
    namespace NodeJS {
        interface Global {
            prisma: PrismaClient;
        }
    }
}

export default function () {
   if (process.env.NODE_ENV === 'production')
      return new PrismaClient()

   if (!global.prisma) global.prisma = new PrismaClient()

   return global.prisma
}
```

**Comment by jaredjj3** [comment]
The solution didn't map 1:1 for me because I'm Prisma to build a CLI/Discord bot. When the bot detects a command, it will spawn a child process and run the command. A new db connection _can_ be created every time a command is run. A command is analogous to an HTTP request like some of the other example solutions.

There are some edge cases where the process can be killed before the command finished running. Before my cleanup solution, the db connections would accumulate and I would encounter the `FATAL: sorry, too many clients already` errors`.

This is what I did to solve my issue.

_onCleanup.ts_
```ts
type CleanupCallback = () => Promise<void>;

export const onCleanup = (() => {
  const callbacks = new Array<CleanupCallback>();

  const cleanup = async () => {
    await Promise.all(callbacks.map((callback) => callback()));
  };

  // these are the signals I needed to listen to, in order to handle
  // user interrupts or system process terminations
  process.on('SIGTERM', cleanup);
  process.on('SIGINT', cleanup);

  return (callback?: CleanupCallback): CleanupCallback => {
    if (callback) {
      callbacks.push(callback);
    }
    return cleanup;
  };
})();
```

And then I'm creating my Prisma client like so:

_inversify.config.ts_
```ts
const prisma = new PrismaClient();
onCleanup(async () => {
  logger.info('disconnecting prisma');
  await prisma.$disconnect();
});
container.bind<PrismaClient>(TYPES.PrismaClient).toConstantValue(prisma);
```

Whenever I inject the Prisma client, it is guaranteed to use the single Prisma client (per process).  I've put measures in place to ensure `cleanup()` is called whenever a command finishes running or errors out.

I thought this was worth sharing because I was encountering the issue when I'm already using a singleton in my code. The key deviation from others is that I have a singleton _per process_. This was not an issue with Prisma.

**Comment by yassinebridi** [comment]
@pantharshit00 I'm using NestJS with the official example provided here:
https://www.prisma.io/nestjs
I get the same problem.
I wonder if there is a trick to debug this, to see how many connection pools are being created.

**Comment by hmiiro** [comment]
@yassinebridi  for nestjs you should be sorted if you write your prisma service as below:


import { Injectable, OnModuleInit, OnModuleDestroy } from '@nestjs/common';
import { PrismaClient } from '@prisma/client';

@Injectable()
export class PrismaService extends PrismaClient
  implements OnModuleInit, OnModuleDestroy {
  constructor() {
    super();
  }
  async onModuleInit() {
    await this.$connect();
  }

  async onModuleDestroy() {
    await this.$disconnect();
  }
}

**Comment by yassinebridi** [comment]
Yep, that's what this example suggests:
prisma.io/nestjs
still, i get that error.

**Comment by juanzgc** [comment]
Will this issue be re-opened? I'm still facing the same issue and it sees others are too. However, the issue remains closed :/

**Comment by glenmcrae** [comment]
Same issue. Ended up switching to Hasura.

**Comment by yassinebridi** [comment]
To get back at this, i switched my server from Heroku to Digitalocean, without any additional configuration. It seems that the problem doesn't exist anymore.

**Comment by macrozone** [comment]
i updated the issue with the solution that solved the problem in my case. If you still have this problem in a different case, I suggest to open a new issue instead of commenting here.

**Comment by bencun** [comment]
Figured out I might add something this to the issue as well, I resolve this problem during development in NestJS by running `pkill -f query-engine-darwin` since it's the native code running in the background keeping the connection open.

This really shouldn't happen, those processes should end for good **at least** when the parent Node process is killed so I still see this as a **MAJOR** issue regardless of the caching workaround.

I think this issue is worth reopening since the current implementation means that when your app crashes completely and then relaunches the Prisma instance caching you have in place will work - but the old connections will still be active. Ending the parent Node process should also terminate those native processes as well.

**Comment by bencun** [comment]
I resolve this in my NestJS app by running `pkill -f query-engine-darwin`.

The issue is in that the query engine processes keep the connection pool open **even when the parent Node process crashes/is killed**. This means massive memory and connection leaks in production if your application is known to crash and is restored automatically by some other entity. Caching the instance does not resolve those use cases.

I think this issue is very important and needs to be reopened.

**Comment by bencun** [comment]
Understandable, I figured it would be better to open a new issue so my previous comment should be disregarded.

**Comment by emipc** [comment]
A combination of previous responses that worked for me, keeping type safety:

```ts
import { PrismaClient } from "@prisma/client";

declare global {
  namespace NodeJS {
    interface Global {
      prisma: PrismaClient;
    }
  }
}

let prisma: PrismaClient;

if (process.env.NODE_ENV === "production") {
  prisma = new PrismaClient();
} else {
  if (!global.prisma) {
    global.prisma = new PrismaClient();
  }

  prisma = global.prisma;
}

export default prisma;

```

Then just import it with:

```ts
import prisma from "../lib/prisma";
```

**Comment by harshamv** [comment]
> A combination of previous responses that worked for me, keeping type safety:
> 
> ```ts
> import { PrismaClient } from "@prisma/client";
> 
> declare global {
>   namespace NodeJS {
>     interface Global {
>       prisma: PrismaClient;
>     }
>   }
> }
> 
> let prisma: PrismaClient;
> 
> if (process.env.NODE_ENV === "production") {
>   prisma = new PrismaClient();
> } else {
>   if (!global.prisma) {
>     global.prisma = new PrismaClient();
>   }
> 
>   prisma = global.prisma;
> }
> 
> export default prisma;
> ```
> 
> Then just import it with:
> 
> ```ts
> import prisma from "../lib/prisma";
> ```

@emipc how can I use this with out typescript syntax?

**Comment by elie222** [comment]
@harshamv there are examples above that do the same without TS.

Remove:
`: PrismaClient`
and
```
declare global {
  namespace NodeJS {
    interface Global {
      prisma: PrismaClient;
    }
  }
}
```

**Comment by harshamv** [comment]
@elie222 thanks a lot

**Comment by windantara** [comment]
> This is a solution I came across, I don't see it posted here but hopefully it helps.
> 
> ```ts
> import { PrismaClient } from '@prisma/client'
> 
> const prismaClientPropertyName = `__prevent-name-collision__prisma`
> type GlobalThisWithPrismaClient = typeof globalThis & {
>   [prismaClientPropertyName]: PrismaClient
> }
> 
> const getPrismaClient = () => {
>   if (process.env.NODE_ENV === `production`) {
>     return new PrismaClient()
>   } else {
>     const newGlobalThis = globalThis as GlobalThisWithPrismaClient
>     if (!newGlobalThis[prismaClientPropertyName]) {
>       newGlobalThis[prismaClientPropertyName] = new PrismaClient()
>     }
>     return newGlobalThis[prismaClientPropertyName]
>   }
> }
> 
> const prisma = getPrismaClient()
> 
> export default prisma
> ```
> 
> @rajuashok this should solve your issue of not being able to get type-completion with the above approach.

you save my life

**Comment by Sytten** [comment]
@mattgabor 
This means you have exhausted the number of connections that your database can accept. You first have to grasp what this means. In postgres and other traditional DBMS, each connection takes either a thread or a process thus consuming quite a bit of memory. This means the database system/VM has a maximum number of concurrent connections it can hold and its materialized by a setting (`max_connections` in postgres). Each instance of your service will open a few connections to the database to improve concurrency thus throughput (otherwise you would only be able to do one DB request at a time, even if you server had hundreds of concurrent incoming requests).

Thus in the following cases, it is possible to "exhaust" the database connections (this is easier to do with postgres since the typical number of connection is around 100):
1. Too many concurrent instances of the service running
2. Each instance holding too many connections in their pool (configured in prisma with the connection params or defaults to NUMBER CORES + 1)
3. Prisma pool still being buggy since the connection reaper is not done properly (I reported the problem a long time ago) so the engine keeps more connections then necessary

Those issues are particularly acute when doing rolling updates in production as the number of instances is 2x for a short period of time. My suggestion is usually to modify your database setting to increase the max number of connections (check current level with `SHOW max_connections;` in postgres) to 2x your normal level (so if you have 100, you set it to 200) so that the rolling update can work. The alternative is to use a connection pooler like pgbouncer for postgres, but that comes with a lot of caveats in prisma (it is not super well supported from my experience).

Hope this helps

**Comment by mattgabor** [comment]
> @mattgabor 
> 
> This means you have exhausted the number of connections that your database can accept. You first have to grasp what this means. In postgres and other traditional DBMS, each connection takes either a thread or a process thus consuming quite a bit of memory. This means the database system/VM has a maximum number of concurrent connections it can hold and its materialized by a setting (`max_connections` in postgres). Each instance of your service will open a few connections to the database to improve concurrency thus throughput (otherwise you would only be able to do one DB request at a time, even if you server had hundreds of concurrent incoming requests).
> 
> 
> 
> Thus in the following cases, it is possible to "exhaust" the database connections (this is easier to do with postgres since the typical number of connection is around 100):
> 
> 1. Too many concurrent instances of the service running
> 
> 2. Each instance holding too many connections in their pool (configured in prisma with the connection params or defaults to NUMBER CORES + 1)
> 
> 3. Prisma pool still being buggy since the connection reaper is not done properly (I reported the problem a long time ago) so the engine keeps more connections then necessary
> 
> 
> 
> Those issues are particularly acute when doing rolling updates in production as the number of instances is 2x for a short period of time. My suggestion is usually to modify your database setting to increase the max number of connections (check current level with `SHOW max_connections;` in postgres) to 2x your normal level (so if you have 100, you set it to 200) so that the rolling update can work. The alternative is to use a connection pooler like pgbouncer for postgres, but that comes with a lot of caveats in prisma (it is not super well supported from my experience).
> 
> 
> 
> Hope this helps

Thanks for the reply and suggestions! After researching it more it looks like the heroku hobby tier I am on only allows for 20 connections max so that may be too few for the setup I have (next.js + prisma). Will look into it.

**Comment by therobhrt** [comment]
Has anyone encountered an issue when Next.js build some number of static pages during the build, which require DB calls, and due to (assumed) high parallelization of these static page builds, there are multiple DB connections spawned simultaneously, tightly correlated with the number of pages being built.

The solution for hot reload does not seem to help in such case, as we always had it in our codebase.

Basically if we pre-build say 10 pages into static, through the build process the nextjs builder opens 10 connections to DB. The bigger problem is that these connections seem to stale, and it's unclear where to edit the code to force terminate them.

**Comment by InternetRamen** [comment]
```ts
import { PrismaClient } from "@prisma/client";

declare global {
    var prisma: PrismaClient;
}

let prisma: PrismaClient;

if (process.env.NODE_ENV === "production") {
    prisma = new PrismaClient();
} else {
    if (!global.prisma) {
        global.prisma = new PrismaClient();
    }

    prisma = global.prisma;
}

export default prisma;
```
A riff on some of the type safety declarations. https://stackoverflow.com/questions/68481686/type-typeof-globalthis-has-no-index-signature

**Comment by vinicius-pimenta** [comment]
In my integration tests, I used the parameter `connection_limit=0` and it stopped the error.

**Comment by josethz00** [comment]
> In my integration tests, I used the parameter `connection_limit=0` and it stopped the error "too many clients already". Setting this to 0 turns off connection limits

it worked for me also, but I didn't understand why. Can you explain me? Makes no sense to me lol

**Comment by LuisPortanel** [comment]
> ```ts
> import { PrismaClient } from "@prisma/client";
> 
> declare global {
>     var prisma: PrismaClient;
> }
> 
> let prisma: PrismaClient;
> 
> if (process.env.NODE_ENV === "production") {
>     prisma = new PrismaClient();
> } else {
>     if (!global.prisma) {
>         global.prisma = new PrismaClient();
>     }
> 
>     prisma = global.prisma;
> }
> 
> export default prisma;
> ```
> 
> A riff on some of the type safety declarations. https://stackoverflow.com/questions/68481686/type-typeof-globalthis-has-no-index-signature

This worked beautifully, except that it has as side effect that if you forget to make your import of prisma, VS Code won't trigger any linter error of missing import during development.

For example, if a controller retrieves prisma.user info and misses the import it will still work like this:
```
// import prisma from '@prisma'  <- If we comment this, it will still work during development
const users = await prisma.users.findMany()
return users
```
But during build time, it will trigger the console error: `ReferenceError: prisma is not defined`

I've found a solution, we just need to use another name for the PrismaClient that we store in the global scope. That way, VS Code won't detect that prisma is defined and we can avoid the "too many clients error":

```
import { PrismaClient } from '@prisma/client'

declare global {
  // eslint-disable-next-line no-var
  var globalPrisma: PrismaClient  // <- Name it whatever you want, except "prisma".
}

let prisma: PrismaClient

if (process.env.NODE_ENV === 'production') {
  prisma = new PrismaClient()
} else {
  if (!global.globalPrisma) {
    global.globalPrisma = new PrismaClient()
  }

  prisma = global.globalPrisma
}

export default prisma

```

Hope it helps anyone :+1:
